kernel/src/fs/notify/inotify/file.rs

```diff
@@ -0,0 +1,485 @@
+            return;
+        }
+        let new_event: Arc<dyn FsnotifyEvent> = Arc::new(InotifyEvent::new(
+            mask,
+            wd,
+            0,
+            (name.len() + 1) as u32,
+            name.clone(),
+        ));
+
+        // Try to merge with the last event if possible
+        let mut notifications = self.notifications.write();
+        let mut merged = false;
+        if let Some(last_event) = notifications.last() {
+            if let Some(last_inotify_event) =
+                (last_event.as_ref() as &dyn Any).downcast_ref::<InotifyEvent>()
+            {
+                // Downcast new_event for comparison
+                if let Some(new_inotify_event) =
+                    (new_event.as_ref() as &dyn Any).downcast_ref::<InotifyEvent>()
+                {
+                    if can_merge_events(last_inotify_event, new_inotify_event) {
+                        // Replace the last event with the new one
+                        notifications.pop();
+                        notifications.push(new_event.clone());
+                        merged = true;
+                    }
+                }
+            }
+        }
+        if !merged {
+            notifications.push(new_event.clone());
+        }
+        drop(notifications);
+        // New or merged event makes the file readable
+        self.pollee.notify(IoEvents::IN);
+    }
+
+    fn pop_event(&self) -> Option<Arc<dyn FsnotifyEvent>> {
+        let mut notifications = self.notifications.write();
+        if notifications.is_empty() {
+            None
+        } else {
+            Some(notifications.remove(0))
+        }
+    }
+
+    fn get_all_event_size(&self) -> usize {
+        self.notifications
+            .read()
+            .iter()
+            .map(|event| event.get_size())
+            .sum()
+    }
+
+    /// Free fsnotify mark of inotify file
+    fn free_mark(&self, mark: &Arc<dyn FsnotifyMark>) {
+        // Send the IN_IGNORED event representing the mark is removed
+        mark.fsnotify_group()
+            .send_event(mark, InotifyMask::IN_IGNORED.bits(), String::new());
+        let wd = mark.downcast_ref::<InotifyMark>().unwrap().wd();
+        self.remove_fsnotify_mark(wd);
+    }
+}
+
+impl Pollable for InotifyFile {
+    fn poll(&self, mask: IoEvents, poller: Option<&mut PollHandle>) -> IoEvents {
+        self.pollee.poll_with(mask, poller, || {
+            if self.get_all_event_size() > 0 {
+                IoEvents::IN
+            } else {
+                IoEvents::empty()
+            }
+        })
+    }
+}
+
+impl FileLike for InotifyFile {
+    fn read(&self, writer: &mut VmWriter) -> Result<usize> {
+        if self.flags.contains(InotifyFlags::IN_NONBLOCK) && self.get_all_event_size() == 0 {
+            return_errno_with_message!(Errno::EAGAIN, "non-blocking read");
+        }
+
+        let mut size = 0;
+        let mut consumed_events = 0;
+        loop {
+            let event = match self.pop_event() {
```

Comment:
> If no events have so far
       occurred, then, assuming a blocking file descriptor, [read(2)](https://man7.org/linux/man-pages/man2/read.2.html) will
       block until at least one event occurs (unless interrupted by a
       signal, in which case the call fails with the error EINTR; see
       [signal(7)](https://man7.org/linux/man-pages/man7/signal.7.html)).

This `read` method does not implement blocking behaviors at all.